File: bindings/typescript/src/alerts.ts
/**
 * The operator alerts inbox view — one terminal-failure alert enumerated by
 * `ListAlerts` (W1a-2). A read-only projection of the journal's TERMINAL `Failed`
 * facts (dead-letters + worker-reported terminal failures) into the gateway's
 * rebuildable-to-empty `alerts.db` read-cache — DISPLAY/TRIAGE-READ ONLY, never
 * truth, never identity, never a digest input. Serve-path admission refusals
 * write nothing to the journal, so they are not in this inbox. Kept in its own
 * module (the telemetry.ts module-per-concern precedent).
 *
 * SN-8: `alertId`/`moteId` are server-derived; the SDK only *encodes* the bytes
 * to hex. The triage LIFECYCLE (acknowledge/resolve), the alert-rule engine, and
 * notifications are a Cloud capability (D156/D129) — there is no mutate method.
 */

import type { AlertSummary as PbAlertSummary } from "./gen/kortecx/v1/gateway_pb.js";
import { encode } from "./hexids.js";

/** One terminal-failure alert: the failed Mote + its watermark-attributed run
 *  (`instanceId` is `""` when unattributed — all-zero on the wire), the failure
 *  class, a display severity (`"error"` | `"refused"`), and the `Failed` fact's
 *  `seq` (the deep-link cursor + pagination). `alertId` is server-derived +
 *  re-fold-stable. */
export class AlertSummary {
  constructor(
    readonly alertId: string,
    readonly moteId: string,
    readonly instanceId: string,
    readonly reasonClass: string,
    readonly reasonCode: number,
    readonly severity: string,
    readonly seq: number,
    readonly createdUnixMs: number,
  ) {}

  static fromProto(a: PbAlertSummary): AlertSummary {
    return new AlertSummary(
      encode(a.alertId),
      encode(a.moteId),
      // All-zero (or empty) = unattributed → "" (the telemetry/global-tail convention).
      a.instanceId.some((b) => b !== 0) ? encode(a.instanceId) : "",
      a.reasonClass,
      a.reasonCode,
      a.severity,
      Number(a.seq),
      Number(a.createdUnixMs),
    );
  }

  /** A plain snake_case object (stable wire-shaped serialization for UIs/logs). */
  toJSON() {
    return {
      alert_id: this.alertId,
      mote_id: this.moteId,
      instance_id: this.instanceId,
      reason_class: this.reasonClass,
      reason_code: this.reasonCode,
      severity: this.severity,
      seq: this.seq,
      created_unix_ms: this.createdUnixMs,
    };
  }
}

/** One page of {@link AlertSummary} (newest-first) plus the `hasMore` cursor flag. */
export interface AlertsPage {
  readonly alerts: AlertSummary[];
  readonly hasMore: boolean;
}

File: bindings/python/src/kortecx/alerts.py
"""Operator alerts inbox view — one ``AlertSummary`` enumerated by ``ListAlerts``
(W1a-2).

A read-only projection of the journal's TERMINAL ``Failed`` facts (dead-letters +
worker-reported terminal failures) into a rebuildable-to-empty ``alerts.db``
read-cache — display/triage-read ONLY, never truth, never identity, never a
digest input. Serve-path admission refusals write nothing to the journal, so
they are not in this inbox. Kept in its own module (the telemetry.py /
module-per-concern precedent). SN-8: ``alert_id``/``mote_id`` are server-derived;
the SDK only hex-encodes the bytes.

The triage LIFECYCLE (acknowledge/resolve), the alert-rule engine, and
notifications are a Cloud capability (D156/D129) — there is no mutate method here.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import List

from . import hexids
from .v1 import gateway_pb2 as _g


@dataclass(frozen=True)
class AlertSummary:
    """One terminal-failure alert: hex ids + the failure class + a display
    severity + the ``Failed`` fact's journal seq (the deep-link cursor +
    pagination). ``instance_id`` is registration-watermark attribution — all-zero
    hex when unattributed."""

    alert_id: str  # hex (server-derived, re-fold-stable)
    mote_id: str  # hex (the failed Mote; deep-link target)
    instance_id: str  # hex (all-zero = unattributed)
    reason_class: str  # terminal FailureReason wire token (e.g. "dead_lettered")
    reason_code: int  # numeric FailureReason discriminant (0-8)
    severity: str  # closed display vocabulary: "error" | "refused"
    seq: int  # the Failed fact's journal seq
    created_unix_ms: int  # audit-only first-folded wall clock (off every hash)

    @classmethod
    def from_proto(cls, a: "_g.AlertSummary") -> "AlertSummary":
        return cls(
            alert_id=hexids.encode(a.alert_id),
            mote_id=hexids.encode(a.mote_id),
            instance_id=hexids.encode(a.instance_id),
            reason_class=a.reason_class,
            reason_code=a.reason_code,
            severity=a.severity,
            seq=a.seq,
            created_unix_ms=a.created_unix_ms,
        )


@dataclass(frozen=True)
class AlertsPage:
    """One newest-first page of :class:`AlertSummary` plus the ``has_more`` flag."""

    alerts: List[AlertSummary]
    has_more: bool
